### PR TITLE
Graph embeddings

### PR DESCRIPTION
**Pull request by lgalke**

### CODE REVIEW

File: aidistillery/graph_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                    g.add_edge(len(g) - 1, paper_idx)
+            else:
+                g.add_edge(paper_idx, tag2idx[tag])
+                if not directed:
+                    g.add_edge(tag2idx[tag], paper_idx)
+
+    idx2tag = {idx: tag for tag, idx in tag2idx.items()}
+    idx2aut = {idx: aut for aut, idx in aut2idx.items()}
+
+    return g, idx2doc, idx2tag, idx2aut
```
Comment (beduffy) [inline aidistillery/graph_embeddings.py]: nice and neat

File: aidistillery/graph_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    print("%d nodes, %d edges." % (g.number_of_nodes(), g.number_of_edges()))
+
+    N = len(g)
+    DAMP = args.damp
+
+    print("N =", N)
+    print("DAMP =", DAMP)
+
+    def pagerank_builtin(g):
+        g.ndata['pv'] = g.ndata['pv'] / g.ndata['deg']
```
Comment (beduffy) [inline aidistillery/graph_embeddings.py]: I've forgotten the details of pagerank :(

File: aidistillery/graph_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                        '--top',
+                        metavar='N',
+                        help="Number of top page-ranked items to return [10]",
+                        default=10,
+                        type=int)
+
+    parser.add_argument('-d', '--damp', help="Damping Factor",
+                        default=.85, type=float)
+
+def gcn_main(args):
```
Comment (beduffy) [inline aidistillery/graph_embeddings.py]: I assume these are not ready yet but you kept them to motivate you to do gcn? :D

File: aidistillery/graph_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    top_authors = top(pv, idx2aut.items(), args.top)
+    for k, (key, score) in enumerate(top_authors):
+        print("Top-{} (pv={:.2E}): {}".format(k+1, score, key))
+
+    print("Top %d hot tags" % args.top)
+    top_tags = top(pv, idx2tag.items(), args.top)
+    for k, (key, score) in enumerate(top_tags):
+        print("Top-{} (pv={:.2E}): {}".format(k+1, score, key))
+
+    print("kthxbye.")
```
Comment (beduffy) [inline aidistillery/graph_embeddings.py]: :rofl:

**Review by beduffy — COMMENTED** [review]
Very good overall, im sure you're gonna be adding to it with gcn. I ran the code, page rank for top papers didn't seem too good though? Ahh, probably because it was just relevant for who had the most authors haha.
Still gonna merge

**Comment by lgalke** [comment]
Yea, at the moment top papers is heavily biased by #authors since that leads to high probability of going from one of all the authors to the paper. Citation data and maybe  twitter data will make the top-papers more insightful.

**Pull request merged**